File: crates/deepsec-scanner/src/walker.rs
use ignore::WalkBuilder;
use std::path::{Path, PathBuf};

/// Directory names we always skip while walking a project. Mirrors the
/// `IGNORE_DIRS` constant in the TS scanner.
pub const IGNORE_DIRS: &[&str] = &[
    "node_modules",
    ".git",
    ".hg",
    ".svn",
    "dist",
    "build",
    "out",
    ".next",
    ".nuxt",
    ".turbo",
    ".cache",
    "coverage",
    "__pycache__",
    ".pytest_cache",
    ".mypy_cache",
    "target",
    "vendor",
    ".terraform",
    ".venv",
    "venv",
    ".tox",
    "deps",
    "Pods",
    ".idea",
    ".vscode",
];

const SKIP_EXTENSIONS: &[&str] = &[
    "md", "lock", "log", "min.js", "map", "ico", "png", "jpg", "jpeg", "gif", "webp", "svg",
    "pdf", "zip", "tar", "gz", "exe", "dll", "so", "dylib", "woff", "woff2", "ttf",
];

/// Walk the project, returning relative file paths. Honors .gitignore.
/// Files matching any [`SKIP_EXTENSIONS`] are dropped — they're noise
/// for source-level analysis.
pub fn walk_project(root: &Path) -> Vec<PathBuf> {
    let mut out = Vec::new();
    let mut builder = WalkBuilder::new(root);
    builder
        .hidden(false)
        .git_ignore(true)
        .git_global(true)
        .git_exclude(true)
        .follow_links(false);
    for d in IGNORE_DIRS {
        builder.add_custom_ignore_filename(d);
    }
    let walker = builder
        .filter_entry(|entry| {
            if let Some(name) = entry.file_name().to_str() {
                if entry.depth() > 0 && IGNORE_DIRS.contains(&name) {
                    return false;
                }
            }
            true
        })
        .build();

    for result in walker {
        let Ok(entry) = result else { continue };
        let path = entry.path();
        if !entry.file_type().map(|f| f.is_file()).unwrap_or(false) {
            continue;
        }
        if let Some(ext) = path.extension().and_then(|e| e.to_str()) {
            if SKIP_EXTENSIONS.iter().any(|s| s.eq_ignore_ascii_case(ext)) {
                continue;
            }
        }
        // skip very large files (>1 MiB) — the AI step caps file size
        // and matchers gain nothing from minified bundles.
        if let Ok(meta) = entry.metadata() {
            if meta.len() > 1_048_576 {
                continue;
            }
        }
        let Ok(rel) = path.strip_prefix(root) else {
            continue;
        };
        if rel.as_os_str().is_empty() {
            continue;
        }
        out.push(rel.to_path_buf());
    }
    out
}

File: internal/scanner/walker.go
package scanner

import (
	"io/fs"
	"os"
	"path/filepath"
	"strings"
)

// IgnoreDirs are directory names always skipped during the walk.
var IgnoreDirs = map[string]struct{}{
	"node_modules":  {},
	".git":          {},
	".hg":           {},
	".svn":          {},
	"dist":          {},
	"build":         {},
	"out":           {},
	".next":         {},
	".nuxt":         {},
	".turbo":        {},
	".cache":        {},
	"coverage":      {},
	"__pycache__":   {},
	".pytest_cache": {},
	".mypy_cache":   {},
	"target":        {},
	"vendor":        {},
	".terraform":    {},
	".venv":         {},
	"venv":          {},
	".tox":          {},
	"deps":          {},
	"Pods":          {},
	".idea":         {},
	".vscode":       {},
}

// skipExtensions are extensions of files that source-level analysis
// gains nothing from (markdown, binaries, generated bundles).
var skipExtensions = map[string]struct{}{
	".md":     {},
	".lock":   {},
	".log":    {},
	".map":    {},
	".ico":    {},
	".png":    {},
	".jpg":    {},
	".jpeg":   {},
	".gif":    {},
	".webp":   {},
	".svg":    {},
	".pdf":    {},
	".zip":    {},
	".tar":    {},
	".gz":     {},
	".exe":    {},
	".dll":    {},
	".so":     {},
	".dylib":  {},
	".woff":   {},
	".woff2":  {},
	".ttf":    {},
	".min.js": {},
}

const maxFileSize = 1 << 20 // 1 MiB

// WalkProject returns the project-relative paths of every file under
// `root` that's a candidate for matching. Skips IgnoreDirs subtrees,
// skipExtensions, files over 1 MiB, and honors .gitignore.
func WalkProject(root string) ([]string, error) {
	ignorer := loadGitignore(root)
	out := make([]string, 0)
	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if path == root {
			return nil
		}
		name := d.Name()
		if d.IsDir() {
			if _, drop := IgnoreDirs[name]; drop {
				return fs.SkipDir
			}
			return nil
		}
		rel, err := filepath.Rel(root, path)
		if err != nil {
			return nil
		}
		rel = filepath.ToSlash(rel)
		if ignorer != nil && ignorer.MatchesPath(rel) {
			return nil
		}
		if _, drop := skipExtensions[strings.ToLower(filepath.Ext(name))]; drop {
			return nil
		}
		// also drop the .min.js suffix specifically (Ext only catches `.js`)
		if strings.HasSuffix(strings.ToLower(name), ".min.js") {
			return nil
		}
		fi, err := os.Stat(path)
		if err != nil || fi.Size() > maxFileSize {
			return nil
		}
		out = append(out, rel)
		return nil
	})
	return out, err
}
